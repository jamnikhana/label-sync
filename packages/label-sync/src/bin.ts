#!/usr/bin/env node

import meow from 'meow'
import { manage } from './manager'

const cli = meow(
  `
Usage
  $ labels

Options
  --config Path to your configuration file.
  --dryrun Simulate sync to get information about potential changes.

Examples
  $ labels --config labels.config.js
`,
  {
    flags: {
      config: {
        type: 'string',
        alias: 'c',
      },
      dryrun: {
        type: 'boolean',
        default: false,
      },
    },
  },
)

/* istanbul ignore next */
if (process.env.NODE_ENV !== 'test') main(cli)

export async function main(cli: meow.Result): Promise<void> {
  manage(cli.flags.config, { dryrun: cli.flags.dryrun }).then(res => {
    if (res.status === 'ok') {
      console.log(res.message)
    } else {
      console.warn(res.message)
    }
  })
}
