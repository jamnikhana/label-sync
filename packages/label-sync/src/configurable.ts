export abstract class Configurable<T> {
  abstract getConfiguration(): T
}
