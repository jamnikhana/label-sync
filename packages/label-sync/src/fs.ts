import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

import { Maybe } from './utils'

const fsReaddir = promisify(fs.readdir)

/**
 * Looks up for a folder starting with the current directory and
 * going up until it reaches the root folder or finds the folder.
 *
 * @param dir
 * @param name
 */
export async function findFolderUp(
  dir: string,
  name: string,
): Promise<Maybe<string>> {
  return findUp(dir, p => path.dirname(p) === name)
}

/**
 * Traverses file system up to find a specific directory that contains
 * a matching pattern.
 *
 * @param dir
 * @param pattern
 */
export async function findUp(
  dir: string,
  pattern: (path: string) => boolean,
): Promise<Maybe<string>> {
  switch (path.normalize(dir)) {
    /* End case: we reached the root. */
    case '/': {
      return null
    }
    /* Recursive case. */
    default: {
      const elements = await fsReaddir(dir)
      const includes = elements.some(pattern)

      if (includes) return dir
      else return findUp(path.join(dir, '../'), pattern)
    }
  }
}
