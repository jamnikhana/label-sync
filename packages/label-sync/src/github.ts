import Octokit from '@octokit/rest'

export interface GithubRepository {
  id: number
  node_id: string
  name: string
  full_name: string
}

/**
 * Fetches all repositories under authenticated user.
 */
export async function getRepositories(
  client: Octokit,
  page: number = 1,
): Promise<GithubRepository[]> {
  const size = 100

  return client.repos
    .list({
      page: page,
      per_page: size,
    })
    .then(async res => {
      if (res.data.length < size) {
        return res.data
      } else {
        const remainingRepositories = await getRepositories(client, page + 1)
        return [...res.data, ...remainingRepositories]
      }
    })
}
