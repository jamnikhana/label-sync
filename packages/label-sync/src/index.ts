import ls from '@label-sync/core'
ls.core
