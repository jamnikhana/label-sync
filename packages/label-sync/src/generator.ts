import fs from 'fs'
import path from 'path'
import { stringify } from 'yaml'

import * as constants from '@labelsync/core/src/constants'

import { Either, right, left, maybe } from './utils'
import { findFolderUp } from './fs'

/**
 * Configuration represents an entire configuration for all
 * LabelSync tools that an organisation is using.
 */
export type LSConfiguration = {
  repos: Map<LSRepositoryName, LSRepository>
}

/**
 * Repository represents a single Github repository.
 * When configured as `strict` it will delete any surplus of labels
 * in the repository.
 */
export type LSRepository = {
  strict: boolean
  labels: Map<LSLabelName, LSLabel>
}
export type LSRepositoryName = string

/**
 * Label represents the central unit of LabelSync. Each label
 * can have multiple siblings that are meaningfully related to
 * a label itself, multiple hooks that trigger different actions.
 */
export type LSLabel = LSLabelDefinition | LSLabelColor
export type LSLabelDefinition = {
  description?: string
  color: string
  siblings?: LSSibling[]
  hooks?: LSHook[]
  // oldName?: string -- TODO: renaming
  // perhaps it would be better if renamings were noted in a separate file
  // or through dialogue to skip the unnecessary field deletion step.
  // * PR message: "I have renamed these labels: - old:new"
}
export type LSLabelName = string
export type LSLabelColor = string

/**
 * Hydrates a Label to a Label definition.
 *
 * @param label
 */
export function lsLabelDefinition(label: LSLabel): LSLabelDefinition {
  switch (typeof label) {
    case 'string': {
      return { color: label }
    }
    case 'object': {
      return label
    }
  }
}

/**
 * Sibling represents a label that LabelSync should add whenever
 * a parent label is assigned to issues or pull request.
 * Siblings can only refer to labels also defined in LabelSync repository
 * configuration.
 */
export type LSSibling = string

/**
 * Represents a label hook. LabelSync triggers a label hook
 * every time a hook is added to an issues or a pull request.
 */
export type LSHook =
  /* webhooks */
  | {
      integration: 'webhook'
      endpoint: string
    }
  /* slack */
  | { integration: 'slack'; action: 'notify'; user: string }
  /* pull requests */
  | { integration: 'pr'; action: 'merge' }
  | { integration: 'pr'; action: 'close' }

/**
 * Validates the configuration file and saves its YAML version
 * to labelsync configuration file path.
 *
 * @param config
 */
export async function save(
  config: LSConfiguration,
  dir: string = __dirname,
  logger: (...args: any[]) => any = console.error,
) {
  /* Loads the configuration. */
  const yaml = generateYAMLConfigurationFromTS(config)

  /* Processes the configuration.  */
  await yaml.either({
    left: errors => {
      logger(reportValidationErrors(errors))
    },
    right: async configuration => {
      /* Finds the location. */
      const projectRoot = await findFolderUp(dir, '.git')
      const labelSyncConfigurationFilePath = path.join(
        maybe(dir, projectRoot),
        constants.labelSyncConfigurationFilePath,
      )

      /* Commit configuration. */
      fs.writeFileSync(labelSyncConfigurationFilePath, configuration)
    },
  })
}

/**
 * Converts ValidationErrors to a human readable format.
 *
 * @param errors
 */
function reportValidationErrors(errors: ValidationError[]): string {
  return `many errors ${errors.length}`
}

/**
 * Turns a TypeScript specification into a label-sync readable
 * YAML configuration.
 *
 * @param values
 */
export function generateYAMLConfigurationFromTS(
  config: LSConfiguration,
): Either<ValidationError[], string> {
  const { errors, configuration } = validateConfiguration(config)
  if (errors) {
    return left(errors)
  }
  return right(stringify(configuration))
}

/**
 * Generates a manifest of TS configuration and validates its input.
 * @param values
 */
function validateConfiguration(
  config: LSConfiguration,
): {
  errors: ValidationError[]
  configuration: LSConfiguration
} {
  /* Helper functions */
  const repos = Object.keys(config.repos)
  const repo = (name: string): LSRepository => config.repos.get(name)!

  /* Validator */
  const errors = repos.reduce<ValidationError[]>((acc, repoName) => {
    const repoConfig = repo(repoName)
    const labels = Object.keys(repoConfig.labels)
    const label = (name: string): LSLabelDefinition =>
      lsLabelDefinition(repoConfig.labels.get(name)!)

    /* Check whether repository configuration defines all siblings. */
    const labelsWithMissingSiblings = labels.filterMap<SiblingsValidationError>(
      labelName => {
        const labelConfig = label(labelName)
        const missingSiblings = maybe([], labelConfig.siblings).filter(
          sibling => labels.some(label => label === sibling),
        )

        if (missingSiblings.length > 0) {
          return {
            repository: repoName,
            label: labelName,
            siblings: missingSiblings,
          }
        } else {
          return null
        }
      },
    )

    return acc.concat(labelsWithMissingSiblings)
  }, [])

  /* Process */

  return { errors: errors, configuration: config }
}

export type ValidationError = SiblingsValidationError

export type SiblingsValidationError = {
  repository: string
  label: string
  siblings: LSSibling[]
}
