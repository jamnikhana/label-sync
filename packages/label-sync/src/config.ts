import * as path from 'path'
import * as fs from 'fs'
import * as Ajv from 'ajv'
import * as Octokit from '@octokit/rest'

import { Config as CoreConfig } from 'label-sync-core'
import {
  RepositoryConfig as CoreRepositoryConfig,
  LabelConfig,
} from 'label-sync-core/dist/labels'

import { getRepositories, GithubRepository } from './github'

/** Schema */

import schema = require('./schema.json')

const ajv = new Ajv()
const validateSchema = ajv.compile(schema)

/**
 * Labels
 */

interface PublishConfig {
  branch: string
}

type RepositoryConfig =
  | string
  | {
      paths: string
      labels?: { [name: string]: LabelConfig }
      strict?: boolean
    }

interface LabelsConfig {
  strict?: boolean
  labels: { [name: string]: LabelConfig }
  repositories: RepositoryConfig[]
  publish: PublishConfig
}

/**
 *
 * Gets JSON labels configuration from absolute path, and validates schema.
 *
 * @param configPath
 */
export function getGithubLabelsJSONConfiguration(
  configPath: string,
): LabelsConfig | null {
  /** Schema import */
  if (!fs.existsSync(configPath) || !path.isAbsolute(configPath)) {
    return null
  }

  const config = JSON.parse(fs.readFileSync(configPath, 'utf8'))

  /** Validation */
  if (!validateSchema(config)) {
    return null
  }

  return config
}

/**
 *
 * Obtains JavaScript configuration file.
 *
 * @param configPath
 */
export function getGithubLabelsJSConfiguration(
  configPath: string,
): CoreConfig | null {
  /** Schema import */
  if (!fs.existsSync(configPath) || !path.isAbsolute(configPath)) {
    return null
  }

  const config = require(configPath)

  return config
}

/**
 * Transforms Github Labels configuration defined in JSON into a format
 * accepted by github-labels-core.
 *
 * @param config
 */
export async function generateConfigurationFromJSONLabelsConfiguration(
  config: LabelsConfig,
  options: {
    githubToken: string
  },
): Promise<
  { status: 'ok'; config: CoreConfig } | { status: 'err'; message: string }
> {
  /**
   * Authentication
   */

  const client = new Octokit()

  client.authenticate({
    type: 'token',
    token: options.githubToken,
  })

  /** Globals */
  const strict = withDefault(false)(config.strict)
  const commonLabels = config.labels
  const repositoryConfigurations = getRepositoryConfigurations(
    withDefault([] as RepositoryConfig[])(config.repositories),
  )

  try {
    /** Remote repositories */
    const repositories = await getRepositories(client)

    /** Definition is a core-compatible config of Github Labels for repo */
    const definitions = generateRepositoryDefinitions(
      repositoryConfigurations,
      repositories,
      {
        strict: strict,
        labels: commonLabels,
      },
    )

    /** Converts to object of repositories */
    const config = definitions.reduce(
      (acc, config) => ({
        ...acc,
        [config.name]: config.config,
      }),
      {},
    )

    return {
      status: 'ok',
      config: config,
    }
  } catch (err) {
    return {
      status: 'err',
      message: err.message,
    }
  }

  /**
   * Helper functions
   */

  /**
   *
   * Merges repository configurations with available repositories using globs.
   *
   * @param configs
   * @param repos
   */
  function generateRepositoryDefinitions(
    configs: {
      paths: RegExp
      isExact: boolean
      labels?: { [name: string]: LabelConfig }
      strict?: boolean
    }[],
    repos: GithubRepository[],
    globals: { strict: boolean; labels: { [name: string]: LabelConfig } },
  ): { name: string; config: CoreRepositoryConfig }[] {
    const repositories = filterMap(repos, repository => {
      /** Finds every repository definition in configuration. */
      const repositoryConfigs = configs.filter(config =>
        config.paths.test(repository.full_name),
      )

      /** Combines definitions into one. */
      if (repositoryConfigs.length > 0) {
        const combinedLabels = repositoryConfigs.reduce<{
          [label: string]: LabelConfig
        }>(
          (acc, config) => ({
            ...acc,
            ...config.labels,
          }),
          {},
        )

        // TODO: Improve strict definition.
        const combinedStrict = repositoryConfigs.find(
          config => config.isExact && config.strict !== undefined,
        )

        return {
          name: repository.full_name,
          config: {
            strict: withDefault(globals.strict)(
              combinedStrict ? combinedStrict.strict : undefined,
            ),
            labels: {
              ...globals.labels,
              ...combinedLabels,
            },
          },
        }
      } else {
        return null
      }
    })

    return repositories
  }

  /**
   *
   * Hydrates string repositories into a full-blown
   * object configuration and replaces paths with RegExp.
   *
   * @param configurations
   */
  function getRepositoryConfigurations(
    configurations: RepositoryConfig[],
  ): {
    paths: RegExp
    isExact: boolean
    labels?: { [name: string]: LabelConfig }
    strict?: boolean
  }[] {
    const hydratedConigurations = configurations.map(configuration => {
      switch (typeof configuration) {
        case 'string':
          return {
            paths: new RegExp(configuration),
            isExact: configuration.includes('*'),
          }
        case 'object':
          return {
            paths: new RegExp(configuration.paths),
            isExact: configuration.paths.includes('*'),
            labels: configuration.labels,
            strict: configuration.strict,
          }
      }
    })

    return hydratedConigurations
  }
}

/**
 * Utils
 */

/**
 *
 * Maps through the values and removes the ones which returned null.
 *
 * @param xs
 * @param fn
 */
function filterMap<X, Y>(xs: X[], fn: (x: X) => Y | null): Y[] {
  return xs.reduce<Y[]>((acc, x) => {
    const res = fn(x)
    if (res !== null) {
      return [...acc, res]
    } else {
      return acc
    }
  }, [])
}

/**
 *
 * Returns fallback if value is undefined.
 *
 * @param fallback
 */
function withDefault<T>(fallback: T): (value: T | undefined) => T {
  return value => {
    if (value !== undefined) {
      return value
    } else {
      return fallback
    }
  }
}
