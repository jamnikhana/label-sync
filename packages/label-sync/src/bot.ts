import { getGithubBot as getGithubBotCore } from 'label-sync-core'
import { getConfigurationFile } from './config'

/**
 *
 * Generates bot from configuration.
 *
 * @param relativeConfigPath
 */
export async function getGithubBot(relativeConfigPath: string) {
  /* Obtains configuration */

  const config = await getConfigurationFile(relativeConfigPath, {})

  if (config.status === 'err') {
    return { status: 'err', message: config.message }
  }

  const bot = getGithubBotCore(config.config)

  return bot
}
