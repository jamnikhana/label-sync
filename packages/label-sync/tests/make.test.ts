import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

import * as ls from '../src'
import { make } from '../src'

const fsReadFile = promisify(fs.readFile)
const fsUnlink = promisify(fs.unlink)

const config = ls.configuration({
  repositories: {
    'prisma-test-utils': ls.repository({
      strict: false,
      labels: {
        'bug/0-needs-reproduction': ls.label('#ff0022'),
        'bug/1-has-reproduction': ls.label({
          color: '#ff0022',
          description: 'Indicates that an issue has reproduction',
        }),
        'bug/2-bug-confirmed': ls.label('red'),
        'bug/3-fixing': ls.label({
          color: '00ff22',
          description: 'Indicates that we are working on fixing the issue.',
        }),
      },
    }),
    'label-sync': ls.repository({
      strict: false,
      labels: {
        'bug/0-needs-reproduction': ls.label('#ff0022'),
        'bug/1-has-reproduction': ls.label({
          color: '#ff0022',
          description: 'Indicates that an issue has reproduction',
        }),
        'bug/2-bug-confirmed': ls.label('red'),
        'bug/3-fixing': ls.label({
          color: '00ff22',
          description: 'Indicates that we are working on fixing the issue.',
        }),
      },
    }),
  },
})

describe('make', () => {
  test('compiles configuration to path', async () => {
    const yamlPath = path.resolve(__dirname, 'labelsync.yml')

    await make({
      configs: [config],
      outputs: {
        config: yamlPath,
      },
    })

    const file = await fsReadFile(yamlPath, { encoding: 'utf-8' })
    await fsUnlink(yamlPath)

    expect(file).toMatchSnapshot()
  })

  test('compiles configuration to default path', async () => {
    await make(
      {
        configs: [config],
      },
      __dirname,
    )

    const yamlPath = path.resolve(__dirname, 'labelsync.yml')
    const file = await fsReadFile(yamlPath, { encoding: 'utf-8' })
    await fsUnlink(yamlPath)

    expect(file).toMatchSnapshot()
  })
})
