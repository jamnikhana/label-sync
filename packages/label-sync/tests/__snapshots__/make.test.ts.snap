// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`make compiles configuration to path 1`] = `
"repos:
  prisma-test-utils:
    strict: false
    labels:
      bug/0-needs-reproduction:
        color: ff0022
      bug/1-has-reproduction:
        color: ff0022
        description: Indicates that an issue has reproduction
      bug/2-bug-confirmed:
        color: red
      bug/3-fixing:
        color: 00ff22
        description: Indicates that we are working on fixing the issue.
  label-sync:
    strict: false
    labels:
      bug/0-needs-reproduction:
        color: ff0022
      bug/1-has-reproduction:
        color: ff0022
        description: Indicates that an issue has reproduction
      bug/2-bug-confirmed:
        color: red
      bug/3-fixing:
        color: 00ff22
        description: Indicates that we are working on fixing the issue.
"
`;
